gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/ProductDependencyMerger.java

```diff
@@ -82,12 +82,18 @@ public static ProductDependency merge(ProductDependency dep1, ProductDependency
                 .filter(version -> satisfiesMaxVersion(maximumVersion, version))
                 .max(VersionComparator.INSTANCE);
 
+        // Since recommended--and thus, discovered--product dependencies cannot be optional, if we are merging
+        // optional and non-optional dependencies, then the one with `optional = true` must be an explicit product
+        // dependency and should always take precedence.
```

Comment:
This seems odd to me - it kind of makes sense for the current world, but if we ever allow pdeps in jars to be optional then I would expect the opposite behaviour, where merging an optional and non-optional dep results in a non-optional dep.

Instead, to me it makes to do all the merging then make the deps that have been marked as optional actually optional. I think this would look like moving this logic:

https://github.com/palantir/sls-packaging/blob/56695cbb51765dd79e91d43aff3dba9a38cddc4e/gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/tasks/CreateManifestTask.java#L254-L257

below the merging, in fact after that `discoverProductDependencies().forEach` block entirely.

Reply:
Cool, I buy that. In which case this logic should probably be `boolean optional = dep1.getOptional() && dep2.getOptional();` to be conservative (because that makes more sense than always `optional = false` even if both the inputs are optional).